confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/DeleteTimeSeriesProcedure.java

```diff
@@ -0,0 +1,611 @@
+          Map<Integer, TSStatus> sendRequest(
+              TDataNodeLocation dataNodeLocation, List<TConsensusGroupId> consensusGroupIdList) {
+            Map<Integer, TDataNodeLocation> dataNodeLocationMap = new HashMap<>();
+            dataNodeLocationMap.put(dataNodeLocation.getDataNodeId(), dataNodeLocation);
+            DeleteTimeSeriesHandler handler = new DeleteTimeSeriesHandler(dataNodeLocationMap);
+            AsyncDataNodeClientPool.getInstance()
+                .sendAsyncRequestToDataNodeWithRetry(
+                    new TDeleteTimeSeriesReq(consensusGroupIdList, patternTreeBytes), handler);
+            return handler.getDataNodeResponseStatusMap();
+          }
+        };
+    deleteTimeSeriesTask.execute();
+  }
+
+  private Map<TDataNodeLocation, List<TConsensusGroupId>> getLeaderDataNodeRegionGroupMap(
+      Map<TConsensusGroupId, Integer> leaderMap,
+      Map<TConsensusGroupId, TRegionReplicaSet> regionReplicaSetMap) {
+    Map<TDataNodeLocation, List<TConsensusGroupId>> dataNodeConsensusGroupIdMap = new HashMap<>();
+    regionReplicaSetMap.forEach(
+        (consensusGroupId, regionReplicaSet) -> {
+          Integer leaderId = leaderMap.get(consensusGroupId);
+          TDataNodeLocation leaderDataNodeLocation = null;
+          if (leaderId == null || leaderId == -1) {
+            leaderDataNodeLocation = regionReplicaSet.getDataNodeLocations().get(0);
+          } else {
+            for (TDataNodeLocation dataNodeLocation : regionReplicaSet.getDataNodeLocations()) {
+              if (dataNodeLocation.getDataNodeId() == leaderId) {
+                leaderDataNodeLocation = dataNodeLocation;
+                break;
+              }
+            }
+          }
+          dataNodeConsensusGroupIdMap
+              .computeIfAbsent(leaderDataNodeLocation, k -> new ArrayList<>())
+              .add(regionReplicaSet.getRegionId());
+        });
+    return dataNodeConsensusGroupIdMap;
+  }
+
+  private Map<TDataNodeLocation, List<TConsensusGroupId>> getAllReplicaDataNodeRegionGroupMap(
+      Map<TConsensusGroupId, TRegionReplicaSet> regionReplicaSetMap) {
+    Map<TDataNodeLocation, List<TConsensusGroupId>> dataNodeConsensusGroupIdMap = new HashMap<>();
+    regionReplicaSetMap.forEach(
+        (consensusGroupId, regionReplicaSet) -> {
+          for (TDataNodeLocation dataNodeLocation : regionReplicaSet.getDataNodeLocations()) {
+            dataNodeConsensusGroupIdMap
+                .computeIfAbsent(dataNodeLocation, k -> new ArrayList<>())
+                .add(regionReplicaSet.getRegionId());
+          }
+        });
+    return dataNodeConsensusGroupIdMap;
+  }
+
+  @Override
+  protected void rollbackState(
+      ConfigNodeProcedureEnv env, DeleteTimeSeriesState deleteTimeSeriesState)
+      throws IOException, InterruptedException, ProcedureException {
+    RegionTask<TSStatus> rollbackStateTask =
+        new RegionTask<TSStatus>(
+            "roll back schema black list",
+            env,
```

Comment:
In which state is the rollback performed?

Reply:
Since schema black list is constructed before all states, it must be rollback to make related timeseries usable again, when any state failed.